File: frontend/src/services/ai.service.ts
/**
 * Service module for AI-powered features.
 *
 * Provides methods for interacting with the AI mock interview system
 * and resume review functionality. All calls go through the shared
 * Axios instance which handles JWT authentication automatically.
 *
 * @author DevLaunch
 */

import apiClient from '../api/client';
import { AI } from '../api/endpoints';
import type {
  StartInterviewRequest,
  StartInterviewResponse,
  SubmitInterviewRequest,
  SubmitInterviewResponse,
  InterviewHistoryResponse,
  InterviewCategoryStats,
  ResumeReviewRequest,
  ResumeReviewResponse,
  TranscribeResponse,
} from '../types/ai';

/** Derives a Whisper-friendly filename from the recorded blob's MIME type. */
function audioFilename(blob: Blob): string {
  const type = blob.type.toLowerCase();
  if (type.includes('mp4')) {
    return 'answer.mp4';
  }
  if (type.includes('ogg')) {
    return 'answer.ogg';
  }
  if (type.includes('wav')) {
    return 'answer.wav';
  }
  return 'answer.webm';
}

export const aiService = {
  /**
   * Starts a new mock interview session for the given category.
   * The backend generates a set of questions for the user to answer.
   */
  startInterview: (data: StartInterviewRequest) =>
    apiClient
      .post<StartInterviewResponse>(AI.START_INTERVIEW, data)
      .then((res) => res.data),

  /**
   * Subjects the user's answers for a given interview session to
   * AI analysis and returns structured feedback with scores.
   */
  submitInterview: (data: SubmitInterviewRequest) =>
    apiClient
      .post<SubmitInterviewResponse>(AI.SUBMIT_INTERVIEW, data)
      .then((res) => res.data),

  /**
   * Retrieves the authenticated user's past interview history,
   * including scores and metadata for each session.
   */
  getInterviewHistory: () =>
    apiClient
      .get<InterviewHistoryResponse>(AI.INTERVIEW_HISTORY)
      .then((res) => res.data),

  /**
   * Retrieves per-category statistics (bank size, best score, last
   * attempt) for the mock interview landing page.
   */
  getInterviewCategories: () =>
    apiClient
      .get<InterviewCategoryStats[]>(AI.INTERVIEW_CATEGORIES)
      .then((res) => res.data),

  /**
   * Deletes a completed interview session from the user's history.
   */
  deleteInterviewSession: (sessionId: string) =>
    apiClient
      .delete<string>(AI.INTERVIEW_HISTORY_BY_SESSION(sessionId))
      .then((res) => res.data),

  /**
   * Submits a resume for AI-powered review and returns suggestions
   * for improvement, missing keywords, and an overall score.
   */
  reviewResume: (data: ResumeReviewRequest) =>
    apiClient
      .post<ResumeReviewResponse>(AI.REVIEW_RESUME, data)
      .then((res) => res.data),

  /**
   * Transcribes a recorded voice answer through the backend Whisper
   * endpoint. The audio blob is uploaded as multipart form data together
   * with the client-tracked recording duration; the OpenAI API key never
   * leaves the server.
   */
  transcribe: (blob: Blob, durationSeconds: number) => {
    const formData = new FormData();
    formData.append('file', blob, audioFilename(blob));
    formData.append('duration', String(Math.round(durationSeconds)));
    return apiClient
      .post<TranscribeResponse>(AI.TRANSCRIBE, formData, {
        headers: { 'Content-Type': 'multipart/form-data' },
        // Whisper transcription of a several-minute answer can take a while.
        timeout: 120_000,
      })
      .then((res) => res.data);
  },
};

File: backend/src/main/java/com/devlaunch/service/interfaces/AiService.java
package com.devlaunch.service.interfaces;

import com.devlaunch.dto.request.MockInterviewStartRequest;
import com.devlaunch.dto.request.MockInterviewSubmitRequest;
import com.devlaunch.dto.request.ResumeReviewRequest;
import com.devlaunch.dto.response.MockInterviewCategoryResponse;
import com.devlaunch.dto.response.MockInterviewFeedbackResponse;
import com.devlaunch.dto.response.MockInterviewHistoryResponse;
import com.devlaunch.dto.response.MockInterviewStartResponse;
import com.devlaunch.dto.response.ResumeReviewResponse;
import com.devlaunch.dto.response.TranscribeResponse;
import com.devlaunch.exception.ResourceNotFoundException;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

/**
 * Service interface for AI-powered features.
 * <p>
 * Defines the contract for AI-driven operations: resume review (analysing
 * an existing resume owned by the authenticated user) and mock interviews
 * (generating practice questions, evaluating the user's answers, returning
 * their practice history, exposing per-category statistics, and allowing
 * sessions to be removed from the history).
 * </p>
 *
 * @author DevLaunch
 */
public interface AiService {

    /**
     * Reviews an existing resume owned by the currently authenticated user.
     * <p>
     * Loads the identified resume together with all of its sections,
     * delegates the analysis to the configured AI provider (falling back
     * to deterministic analysis when no provider is configured or fails),
     * and returns the structured review result.
     * </p>
     *
     * @param request the review request containing the resume ID and an
     *                optional target role
     * @return the structured resume review result
     * @throws ResourceNotFoundException if the resume is not found or does
     *                                   not belong to the authenticated user
     */
    ResumeReviewResponse reviewResume(ResumeReviewRequest request);

    /**
     * Starts a new mock interview session for the given configuration.
     * <p>
     * Generates a set of interview questions using the configured AI
     * provider (falling back to the deterministic question bank) and
     * returns them together with a session identifier used to submit the
     * answers later.
     * </p>
     *
     * @param request the start request containing the interview category,
     *                difficulty, question length, and timed flag
     * @return the generated questions and session identifier
     */
    MockInterviewStartResponse startMockInterview(MockInterviewStartRequest request);

    /**
     * Submits the answers of a mock interview session for AI evaluation.
     * <p>
     * Evaluates each answer with the configured AI provider (falling back
     * to deterministic heuristics), persists the completed session with
     * its full report to the user's interview history, and returns the
     * structured feedback.
     * </p>
     *
     * @param request the submit request containing the session identifier,
     *                category, configuration, and question/answer pairs
     * @return the structured interview feedback
     */
    MockInterviewFeedbackResponse submitMockInterview(MockInterviewSubmitRequest request);

    /**
     * Retrieves the authenticated user's completed interview history,
     * ordered most recent first, together with the aggregate statistics
     * that power the landing page and analytics views.
     *
     * @return the interview history with summary statistics
     */
    MockInterviewHistoryResponse getMockInterviewHistory();

    /**
     * Retrieves per-category statistics for the landing page: the question
     * bank size together with the authenticated user's practice history
     * for each category.
     *
     * @return the per-category statistics
     */
    List<MockInterviewCategoryResponse> getMockInterviewCategories();

    /**
     * Deletes a completed interview session from the authenticated user's
     * history.
     *
     * @param sessionId the client-generated session identifier to delete
     * @throws ResourceNotFoundException if the session is not found or does
     *                                   not belong to the authenticated user
     */
    void deleteMockInterview(String sessionId);

    /**
     * Transcribes a recorded voice answer through the configured speech-to-
     * text provider (OpenAI Whisper).
     * <p>
     * Validates the uploaded audio and delegates the transcription to the
     * shared Whisper client so the API key never leaves the backend. Returns
     * the transcribed text together with the audio duration used for
     * speaking analytics.
     * </p>
     *
     * @param file                  the recorded audio file
     * @param clientDurationSeconds the recording duration tracked by the
     *                              client, used as a fallback duration
     * @return the transcript and audio duration
     */
    TranscribeResponse transcribe(MultipartFile file, Integer clientDurationSeconds);

}
